File: C++17/백준/Gold/1588. 수열/수열.cc
#include <iostream>
using ll = long long;
using namespace std;

const int rules[4][3] = { {}, {1, 3, 2}, {2, 1, 1}, {2, 3, 2}};
int init, N;
ll L, R;

ll dp[21][4][3];
ll answer[3];

void get_counts(int num, int time, ll start, ll end) {
    if (end < L || start > R) {
        return;
    }

    if (L <= start && end <= R) {
        for (int i = 0; i < 3; i++)
            answer[i] += dp[time][num][i];
        return;
    }

    ll step = (end - start + 1) / 3;
    ll curr_start = start;
    for (const auto& nxt_num: rules[num]) {
        get_counts(nxt_num, time - 1, curr_start, curr_start + step - 1);
        curr_start += step;
    }

    return;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    cin >> init >> L >> R >> N;

    for (int i = 1; i < 4; i++) {
        dp[0][i][i - 1] = 1;
    }

    for (int t = 1; t <= N; t++) {
        for (int num = 1; num < 4; num ++) {
            const auto& r = rules[num];
            for (int i = 0; i < 3; i++) {
                dp[t][num][i] = dp[t - 1][r[0]][i] + dp[t - 1][r[1]][i] + dp[t - 1][r[2]][i];
            }
        }
    }

    ll total_len = dp[N][init][0] + dp[N][init][1] + dp[N][init][2];
    get_counts(init, N, 0, total_len - 1);
    for (int i = 0; i < 3; i++)
        cout << answer[i] << ' ';

    return 0;
}

File: PyPy3/백준/Gold/1588. 수열/수열.py
import sys
input = sys.stdin.readline

init = int(input())
L = int(input())
R = int(input())
N = int(input())

dp = [[[0] * 3 for _ in range(4)] for _ in range(N + 1)]

for num in range(1, 4):
    dp[0][num][num - 1] = 1
    
rules = {1: [1, 3, 2], 2: [2, 1, 1], 3: [2, 3, 2]}

for t in range(1, N + 1):
    for num in range(1, 4):
        r1, r2, r3 = rules[num]
        for i in range(3):
            dp[t][num][i] = dp[t - 1][r1][i] + dp[t - 1][r2][i] + dp[t - 1][r3][i]

def get_counts(num, time, start, end):
    if end < L or start > R:
        return [0, 0, 0]

    if L <= start and end <= R:
        return dp[time][num]
    
    res = [0, 0, 0]
    if time == 0:
        res[num - 1] = 1
        return res
    
    step = (end - start + 1) // 3
    curr_start = start
    for next_num in rules[num]:
        sub = get_counts(next_num, time - 1, curr_start, curr_start + step - 1)
        for i in range(3):
            res[i] += sub[i]
        curr_start += step
    
    return res

total_len = 3 ** N
answer = get_counts(init, N, 0, total_len - 1)
print(*answer)